keyboards/rubi/rubi.c

```diff
@@ -0,0 +1,107 @@
+/* Copyright 2021 gregorio
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation, either version 2 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+#include "rubi.h"
+
```

Comment:
Most of the functions here don't have the user function called, meaning that users cannot control the functionality. 

I would prefer that that the user functions be used here, and set to weak, so they can be overridden.

Reply:
Ah, I forgot to call the user functions. I could just call them before or after the keyboard function, right? For example, I could set the `state` variable for `layer_state_set_kb` function to the returned value of `layer_state_set_user` function, like below:
```
layer_state_t layer_state_set_kb(layer_staate_t state) {
    state = layer_state_set_user(state);
    current_layer = get_highest_layer(state);
    return state;
}
```

Reply:
I choose the same way for the `led_update_kb` function. For the `encoder_update_kb` function, as your suggestion I moved its content to a weak `encoder_update_user` function and call it in the `encoder_update_user` function. Any other things I need to do before I push this changes?

Reply:
That's the best way, for the most part.  

Though, `current_layer` isn't needed.  `layer_state` is a globally available bitmask of the layer state, so it is accessible anywhere.

Reply:
Implemented in https://github.com/qmk/qmk_firmware/pull/12283/commits/9cf41ee489b38ce8e1e07a85ac57f20674a23488.